daemons/controld/controld_cib.c

```diff
@@ -401,57 +401,85 @@ cib_delete_callback(xmlNode *msg, int call_id, int rc, xmlNode *output,
-                           int options)
+controld_node_state_deletion_strings(const char *uname,
+                                     enum controld_section_e section,
+                                     char **xpath, char **desc)
 {
-    cib_t *cib = controld_globals.cib_conn;
-    char *xpath = NULL;
-    char *desc = NULL;
-    int cib_rc = pcmk_ok;
+    const char *desc_pre = NULL;
 
     // Shutdown locks that started before this time are expired
     long long expire = (long long) time(NULL)
                        - controld_globals.shutdown_lock_limit;
 
-    CRM_ASSERT((uname != NULL) && (cib != NULL));
+    // No need to null; we'll set these to something or assert
+    free(*xpath);
+    if (desc != NULL) {
+        free(*desc);
+    }
 
     switch (section) {
         case controld_section_lrm:
-            xpath = crm_strdup_printf(XPATH_NODE_LRM, uname);
-            desc = crm_strdup_printf("resource history for node %s", uname);
+            *xpath = crm_strdup_printf(XPATH_NODE_LRM, uname);
+            desc_pre = "resource history";
             break;
         case controld_section_lrm_unlocked:
-            xpath = crm_strdup_printf(XPATH_NODE_LRM_UNLOCKED,
-                                      uname, uname, expire);
-            desc = crm_strdup_printf("resource history (other than shutdown "
-                                     "locks) for node %s", uname);
+            *xpath = crm_strdup_printf(XPATH_NODE_LRM_UNLOCKED,
+                                       uname, uname, expire);
+            desc_pre = "resource history (other than shutdown locks)";
             break;
         case controld_section_attrs:
-            xpath = crm_strdup_printf(XPATH_NODE_ATTRS, uname);
-            desc = crm_strdup_printf("transient attributes for node %s", uname);
+            *xpath = crm_strdup_printf(XPATH_NODE_ATTRS, uname);
+            desc_pre = "transient attributes";
             break;
         case controld_section_all:
-            xpath = crm_strdup_printf(XPATH_NODE_ALL, uname);
-            desc = crm_strdup_printf("all state for node %s", uname);
+            *xpath = crm_strdup_printf(XPATH_NODE_ALL, uname);
+            desc_pre = "all state";
             break;
         case controld_section_all_unlocked:
-            xpath = crm_strdup_printf(XPATH_NODE_ALL_UNLOCKED,
-                                      uname, uname, expire, uname);
-            desc = crm_strdup_printf("all state (other than shutdown locks) "
-                                     "for node %s", uname);
+            *xpath = crm_strdup_printf(XPATH_NODE_ALL_UNLOCKED,
+                                       uname, uname, expire, uname);
+            desc_pre = "all state (other than shutdown locks)";
             break;
         default:
-            // Should be impossible
-            return;
+            // We called this function incorrectly
+            CRM_ASSERT(false);
```

Comment:
We don't use `CRM_ASSERT(false)` much. IMO since this is daemon code, asserting false here is more convenient than returning an error code and letting the caller assert.

Reply:
That's fine